rfd/0171-database-session-playback.md

```diff
@@ -0,0 +1,376 @@
+}
+
+// DatabaseSessionCommandResult represents the result of a user command. It is
+// expected that for each user command/query there will be a corresponding
+// result.
+message DatabaseSessionCommandResult {
+  // Metadata is a common event metadata.
+  Metadata Metadata = 1;
+  // User is a common user event metadata.
+  UserMetadata User = 2;
+  // SessionMetadata is a common event session metadata.
+  SessionMetadata Session = 3;
+  // Database contains database related metadata.
+  DatabaseMetadata Database = 4;
+
+  oneof Result {
+    // Status of the query execution. It is used when the command doesn't return
+    // any data.
+    Status status = 5;
+    // Data contains the fields and data of the query execution.
+    DatabaseResultData Data = 6;
+  }
+}
+```
+
+#### Examples
+
+* Audit event generated for a query that returns data (`SELET id, name FROM events;`):
+
+```json
+{
+  ...
+  "data": {
+    "fields": [
+      {"name": "id", "type": "INTEGER"},
+      {"name": "name", "type": "STRING"}
+    ],
+    "rows": [
+      ["1", "session.start"],
+      ["2", "session.end"]
+    ]
+  }
+}
+```
+
+* Query that only returns a success message (`UPDATE events SET ...`):
+
+```json
+{
+  ...
+  "status": {
+    "success": true,
+    "message": "3 rows affected"
+  }
+}
+```
+
+* Query that returns an error (`SELECT err;`):
+
+```json
+{
+  ...
+  "status": {
+    "success": false,
+    "error": "ERROR: column \"err\" does not exist"
+  }
+}
+```
+
+### Recording options
+
+Database session recording options will be defined per role. This configuration
+will be under the `options.record_session.db` role option. It will support the
+following values:
+- `on`: Enables session recording, but only queries are recorded. This will
+  mimic the current behavior.
+- `full`: Enables session recording. Queries and responses are recorded.
+- `off`: Disables session recording. Audit events are kept unchanged (start,
+  end, and query events emitted).
+- `recording_only`: Enables recording, but events will not be emitted, they
+  will only be present on recordings. This mode will be the same format as SSH
+  sessions, where the commands and their results are only present on the
+  recording.
+
+The value will be set to `on` by default, keeping the current behavior.
+
+```yaml
+kind: role
+version: v5
+metadata:
+  name: alice
+spec:
+  allow:
+    ...
+  options:
+    record_session:
+      db: on|full|recording_only|off
+```
+
+### Player
+
+To play the session recording, we will rely on the existing player used by SSH
+sessions. This means we don't need to introduce any new player on `tsh` or the
+Web UI. We're going to convert database recording events into `SessionPrint`
+events, which can be rendered by the players.
+
+This conversion will be agnostic to the database protocol. This will make it
+easier to keep the session recording play consistent across protocols.
+
+It will present the session recording events in different text formats:
```

Comment:
Could you add some explanation on the discrepancy between user's terminal session vs replayed session in this `Player` Section? for example `\du` from `psql`

Probably an implementation detail but we should deliver this message when user uses the player for db sessions.

Reply:
I've added a description section on this.